### PR TITLE
Remote Config Refactor Public and Response types

### PR DESCRIPTION
**Pull request by lahirumaramba**
- Refactor `public` and `response` types
- Change the `response` DTOs to simple java beans (with getters and setters)
- Add `protected` `ctors` to `public` types to initialize from the corresponding `response` types

Related to: #446

### CODE REVIEW

File: src/main/java/com/google/firebase/remoteconfig/RemoteConfigParameter.java [hunk 1]
```diff
@@ -122,9 +136,21 @@ ParameterResponse toParameterResponse() {
...
-            conditionalResponseValues);
+    ParameterValueResponse defaultValueResponse = (defaultValue == null) ? null
+            : defaultValue.toParameterValueResponse();
+    return new ParameterResponse()
+            .setDefaultValue(defaultValueResponse)
+            .setDescription(description)
+            .setConditionalValues(conditionalResponseValues);
+  }
+
+  private RemoteConfigParameterValue fromParameterValueResponse(
```
Comment (hiranya911) [inline src/main/java/com/google/firebase/remoteconfig/RemoteConfigParameter.java:147]: I think this ought to be in the `ParameterValue` class. `ParameterValue.fromParameterValueResponse()` makes sense.

File: src/main/java/com/google/firebase/remoteconfig/RemoteConfigParameter.java [hunk 2]
```diff
@@ -45,6 +44,21 @@ public RemoteConfigParameter() {
...
+    checkNotNull(parameterResponse);
+    conditionalValues = new HashMap<>();
+    if (parameterResponse.getConditionalValues() != null) {
+      for (Map.Entry<String, ParameterValueResponse> entry
+              : parameterResponse.getConditionalValues().entrySet()) {
+        conditionalValues.put(entry.getKey(), fromParameterValueResponse(entry.getValue()));
+      }
+    }
+    ParameterValueResponse responseDefaultValue = parameterResponse.getDefaultValue();
+    defaultValue = (responseDefaultValue == null) ? null
```
Comment (hiranya911) [inline src/main/java/com/google/firebase/remoteconfig/RemoteConfigParameter.java:57]: Use `this.` to indicate local members. Particularly useful in constructor implementations.

File: src/main/java/com/google/firebase/remoteconfig/RemoteConfigParameterValue.java [hunk 3]
```diff
@@ -67,7 +67,9 @@ public String getValue() {
 
     @Override
     ParameterValueResponse toParameterValueResponse() {
-      return ParameterValueResponse.ofValue(this.value);
+      return new ParameterValueResponse()
+              .setValue(this.value)
+              .setInAppDefaultValue(null);
```
Comment (hiranya911) [inline src/main/java/com/google/firebase/remoteconfig/RemoteConfigParameterValue.java:72]: Remove redundant `null` assignments.

File: src/main/java/com/google/firebase/remoteconfig/internal/TemplateResponse.java [hunk 4]
```diff
@@ -110,27 +96,22 @@ public RemoteConfigParameter toRemoteConfigParameter() {
...
-    private ParameterValueResponse(String value, Boolean inAppDefaultValue) {
-      this.value = value;
-      this.inAppDefaultValue = inAppDefaultValue;
+    public String getValue() {
+      return value;
     }
 
-    public static ParameterValueResponse ofValue(String value) {
-      return new ParameterValueResponse(value, null);
+    public Boolean isInAppDefaultValue() {
```
Comment (hiranya911) [inline src/main/java/com/google/firebase/remoteconfig/internal/TemplateResponse.java:103]: `is***()` methods should never return null. Instead change this to one of the following:

```java
public Boolean getUseInAppDefault() {
  return this.useInAppDefault;
}

public boolean isUseInAppDefault() {
  return Boolean.TRUE.equals(this.useInAppDefault);
}
```

File: src/main/java/com/google/firebase/remoteconfig/internal/TemplateResponse.java [hunk 5]
```diff
@@ -110,27 +96,22 @@ public RemoteConfigParameter toRemoteConfigParameter() {
     @Key("useInAppDefault")
     private Boolean inAppDefaultValue;
```
Comment (hiranya911) [inline src/main/java/com/google/firebase/remoteconfig/internal/TemplateResponse.java:97]: useInAppDefault

File: src/main/java/com/google/firebase/remoteconfig/internal/TemplateResponse.java [hunk 6]
```diff
@@ -110,27 +96,22 @@ public RemoteConfigParameter toRemoteConfigParameter() {
...
+      this.value = value;
+      return this;
     }
 
-    public RemoteConfigParameterValue toRemoteConfigParameterValue() {
-      if (this.inAppDefaultValue != null && this.inAppDefaultValue) {
-        return RemoteConfigParameterValue.inAppDefault();
-      }
-      return RemoteConfigParameterValue.of(this.value);
+    public ParameterValueResponse setInAppDefaultValue(Boolean inAppDefaultValue) {
```
Comment (hiranya911) [inline src/main/java/com/google/firebase/remoteconfig/internal/TemplateResponse.java:112]: Implement a setter that matches the getter you choose to implement.

**Review by hiranya911 — CHANGES_REQUESTED** [review]
Looks great. The new pattern is working out nicely. Just a few nits to clean up the code.

File: src/main/java/com/google/firebase/remoteconfig/internal/TemplateResponse.java (hunk 4 above)
Comment (lahirumaramba) [reply src/main/java/com/google/firebase/remoteconfig/internal/TemplateResponse.java:103]: Ah good point! Thanks!

File: src/main/java/com/google/firebase/remoteconfig/internal/TemplateResponse.java (hunk 6 above)
Comment (lahirumaramba) [reply src/main/java/com/google/firebase/remoteconfig/internal/TemplateResponse.java:112]: Should the getter take a `boolean`?

File: src/main/java/com/google/firebase/remoteconfig/internal/TemplateResponse.java [hunk 7]
```diff
@@ -108,29 +94,28 @@ public RemoteConfigParameter toRemoteConfigParameter() {
...
 
-    public ParameterValueResponse() {
+    public String getValue() {
+      return value;
     }
 
-    private ParameterValueResponse(String value, Boolean inAppDefaultValue) {
-      this.value = value;
-      this.inAppDefaultValue = inAppDefaultValue;
+    public Boolean getUseInAppDefault() {
```
Comment (hiranya911) [inline src/main/java/com/google/firebase/remoteconfig/internal/TemplateResponse.java:103]: Is this used anywhere? If not lets remove it. And also change the setter to accept `boolean` so it matches the getter.

**Review by hiranya911 — APPROVED** [review]
LGTM with a suggestion.

File: src/main/java/com/google/firebase/remoteconfig/internal/TemplateResponse.java (hunk 7 above)
Comment (lahirumaramba) [reply src/main/java/com/google/firebase/remoteconfig/internal/TemplateResponse.java:103]: Good call. Updated!

**Pull request merged**